components/api_request_helper/api_request_helper.cc

```diff
@@ -277,8 +328,88 @@ void APIRequestHelper::OnDownload(SimpleURLLoaderList::iterator iter,
     }
   }
 
-  url_loaders_.erase(iter);
+  Cancel(iter);
   std::move(callback).Run(path, std::move(headers));
 }
 
+APIRequestHelper::LoaderWrapperHandler::LoaderWrapperHandler(
+    APIRequestHelper* api_request_helper)
+    : api_request_helper_(api_request_helper) {}
+APIRequestHelper::LoaderWrapperHandler::~LoaderWrapperHandler() = default;
+
+void APIRequestHelper::LoaderWrapperHandler::OnDataReceived(
+    base::StringPiece string_piece,
+    base::OnceClosure resume) {
+  // Binding a remote will guarantee the order of the messages being sent, so we
+  // must ensure that it is connected.
+  DCHECK(api_request_helper_->data_decoder_);
+  DCHECK(api_request_helper_->json_parser_.is_connected());
+
+  // We split the string into multiple chunks because there are cases where
+  // multiple chunks are received in a single call.
+  std::vector<std::string> stream_data = base::SplitString(
+      string_piece, "\r\n", base::KEEP_WHITESPACE, base::SPLIT_WANT_NONEMPTY);
+  static constexpr char kDataPrefix[] = "data: {";
+
+  for (const auto& data : stream_data) {
+    if (!base::StartsWith(data, kDataPrefix)) {
+      continue;
+    }
+
+    std::string json = data.substr(strlen(kDataPrefix) - 1);
+
+    api_request_helper_->data_decoder_->ParseJson(
+        json,
+        base::BindOnce(
+            &APIRequestHelper::LoaderWrapperHandler::OnParseJsonFromStream,
+            weak_ptr_factory_.GetWeakPtr()));
+  }
+
+  std::move(resume).Run();
+}
+
+void APIRequestHelper::LoaderWrapperHandler::OnComplete(bool success) {
+  DCHECK(data_completed_callback_);
+  DVLOG(1) << "[[" << __func__ << "]]"
+           << " Response completed\n";
+
+  int response_code = -1;
+
+  if (url_loader_ && url_loader_->ResponseInfo()->headers) {
+    response_code = url_loader_->ResponseInfo()->headers->response_code();
+  }
+
+  std::move(data_completed_callback_).Run(success, response_code);
+  api_request_helper_->Cancel(this);
+}
+
+void APIRequestHelper::LoaderWrapperHandler::OnRetry(
+    base::OnceClosure start_retry) {
+  // Retries are not enabled for these requests.
+  NOTREACHED();
+}
```

Comment:
If we're just going to crash during OnRetry, isn't it better to not override this function, so that the retries (which occasionally happen with a spotty network) go through successfully?

Reply:
You cannot selectively override virtual function unless if there's a default defined in the definition. There's an upstream [example](https://source.chromium.org/chromium/chromium/src/+/main:components/speech/downstream_loader.cc;l=44?q=%22%2F%2F%20Retries%20are%20not%20enabled%20for%20these%20requests.%22) that does the same.

Reply:
retries are enabled